packages/hop-node/src/watchers/xDaiBridgeWatcher.ts

```diff
@@ -101,89 +56,110 @@ class xDaiBridgeWatcher extends BaseWatcher {
+      return
+    }
+    await this.db.transferRoots.update(transferRootHash, {
+      sentConfirmTxAt: Date.now()
+    })
+
     try {
-      const l1Amb = getL1Amb(this.tokenSymbol)
-      const l2Amb = getL2Amb(this.tokenSymbol)
-
-      this.logger.debug(`xDai ${this.tokenSymbol} bridge watcher started`)
-      while (true) {
-        if (!this.started) {
-          return
-        }
-        const blockNumber = await l2Amb.provider.getBlockNumber()
-        const events = await l2Amb.queryFilter(
-          l2Amb.filters.UserRequestForSignature(),
-          blockNumber - 100
-        )
-
-        for (const event of events) {
-          try {
-            const result = await executeExitTx(event, this.tokenSymbol)
-            if (result == null) {
-              continue
-            }
-            const { tx, msgHash } = result
-            this.logger.debug(`executeSignatures. tx hash: ${tx.hash} messageHash: ${msgHash}`)
-          } catch (err) {
-            this.logger.error('tx error:', err.message)
-          }
-        }
-        await wait(10 * 1000)
+      const result = await this.relayXDomainMessage(commitTxHash)
+      if (result == null) {
+        logger.error('no result returned from exit tx')
+        return
       }
+      const { tx } = result
+      const msg = `sent chainId ${this.bridge.chainId} confirmTransferRoot L1 exit tx ${tx.hash}`
+      logger.info(msg)
+      this.notifier.info(msg)
     } catch (err) {
-      this.logger.error('xDai bridge watcher error:', err)
-      this.quit()
+      logger.error(err)
+      throw err
     }
   }
 
-  async handleCommitTxHash (commitTxHash: string, transferRootHash: string, logger: Logger) {
-    const dbTransferRoot = await this.db.transferRoots.getByTransferRootHash(transferRootHash)
-    const destinationChainId = dbTransferRoot?.destinationChainId
-    const l2Amb = getL2Amb(this.tokenSymbol)
+  async relayXDomainMessage (commitTxHash: string) {
+    const token: string = this.tokenSymbol
+    const l1Amb = getL1Amb(token)
+    const l2Amb = getL2Amb(token)
+
+    const sigEvent = await this.getValidSigEvent(commitTxHash)
+    if (!sigEvent) {
+      this.logger.error(`sigEvent not found for ${commitTxHash}`)
+    }
+
+    this.logger.info('found sigEvent event')
+    const message = sigEvent!.args.encodedData
+    if (!message) {
+      this.logger.error(`message not found for ${commitTxHash}`)
+    }
+
+    const msgHash = solidityKeccak256(['bytes'], [message])
+    const id = await l2Amb.numMessagesSigned(msgHash)
+    const alreadyProcessed = await l2Amb.isAlreadyProcessed(id)
+    if (!alreadyProcessed) {
+      this.logger.error(`commit already processed found for ${commitTxHash}`)
+      return
+    }
+
+    const messageId =
+      '0x' +
+      Buffer.from(strip0x(message), 'hex')
+        .slice(0, 32)
+        .toString('hex')
+    const alreadyRelayed = await l1Amb.relayedMessages(messageId)
+    if (alreadyRelayed) {
+      return
```

Comment:
should we log that it's already been relayed here?

Reply:
Updated to throw. Addressed in top level comment below.